wpaclient.lua

```diff
@@ -92,8 +92,18 @@ end
 local network_mt = {__index = {}}
 
 function network_mt.__index:getSignalQuality()
-    -- convert from RSSI to signal quality in range of [0%, 100%].
-    return math.min(math.max((self.signal_level + 100) * 2, 0), 100)
+    -- Make sure the RSSI is in a positive range (hopefully one that's in the ballpark of [0%, 100%]).
+    -- There's no real silver bullet here, as the RSSI is in arbitrary units,
+    -- which means every driver kinda does what it wants with it...
+
+    -- So, at the very least, attempt to detect those that report it as a dBm value, because it'll be negative.
+    if self.signal_level < 0
+        -- Actually hitting 0 dBm is highly unlikely, but, assuming [-100, 0] nonetheless appears to give consistent
+        -- results across the WiFi drivers used on Kobo devices...
+        return self.signal_level + 100
```

Comment:
+100? Is that quite right?

Reply:
Maybe not, but it seems consistent with what the Forma does vs. a dBm value on the H2O and my phone.

Reply:
Seems a bit odd, but oh well.

Reply:
I'm open to better ideas ^^.

Reply:
It's not completely clear to me what kind of actual values are returned. Afaik a realistic range would be from something like -30 dBm (best) down to -80/-90 dBm where you'd probably be SOL but you'd still see the AP.

None of that is a problem for connecting and sorting, but it seems to me that if you take 0 as the highest you can get (which you can't) you'd only show 4 bars even when you're right next to your AP?

Reply:
Well, that's kind of the issue: it's an arbitrary unit, so, we don't know ;p.

That's what my "-0 is probably highly unrealistic" comment alluded to (e.g., at most we might hit -20 if sitting very near the AP); but, doing it this way does get me consistent results across three devices with three different RSSI reporting methods.

And, apparently, on Kobos with a positive RSSI, people reported > 80 values, so I can't really fudge a +30 somewhere without screwing up something else ^^.

----

The strength icons are already hella broken in Nickel itself, so, I prefer sane sorting, mostly consistent scale, and slightly off icons than the alternative ;).

Reply:
And I, in fact, get well below -30 dBm on the H2O if I stick it near the router:

```
eth0      IEEE 802.11-DS  ESSID:"Serenity"  Nickname:""
          Mode:Managed  Frequency:2.472 GHz  Access Point: AA:BB:CC:DD:EE:FF   
          Bit Rate=65 Mb/s   Tx-Power:32 dBm   
          Retry min limit:7   RTS thr:off   Fragment thr:off
          Power Management:off
          Link Quality=5/5  Signal level=-15 dBm  Noise level=-92 dBm
          Rx invalid nwid:0  Rx invalid crypt:0  Rx invalid frag:0
          Tx excessive retries:1  Invalid misc:0   Missed beacon:0
```

Reply:
And that's roughly consistent with the Forma:

```
eth0      IEEE 802.11bgn  ESSID:"Serenity"  Nickname:"<WIFI@REALTEK>"
          Mode:Managed  Frequency:2.472 GHz  Access Point: AA:BB:CC:DD:EE:FF   
          Bit Rate:144.4 Mb/s   Sensitivity:0/0  
          Retry:off   RTS thr:off   Fragment thr:off
          Encryption key:****-****-****-****-****-****-****-****   Security mode:open
          Power Management:off
          Link Quality=87/100  Signal level=79/100  Noise level=0/100
          Rx invalid nwid:0  Rx invalid crypt:0  Rx invalid frag:0
          Tx excessive retries:0  Invalid misc:0   Missed beacon:0
```

Reply:
Perhaps paired with something like https://github.com/NiLuJe/koreader/commit/61c4b4d58a50a1fd3606575eaf3e14d6c2712101 ?

Reply:
> And I, in fact, get well below -30 dBm on the H2O if I stick it near the router:

Is that physically plausible? >_>

Reply:
> Perhaps paired with something like NiLuJe/koreader@500158b ?

Yeah, if the Kobo claims it gets -15 dBm... lol

Reply:
Slightly revamped the logic after taking a look at what NM does... :).